lib/dfe/analytics/fields.rb

```diff
@@ -24,25 +24,21 @@ def self.surplus_fields
       end
 
       def self.diff_model_attributes_against(*lists)
-        Rails.application.eager_load!
-        ActiveRecord::Base.descendants
-          .reject { |model| model.name.include? 'ActiveRecord' } # ignore internal AR classes
-          .reduce({ missing: {}, surplus: {} }) do |diff, next_model|
-            table_name = next_model.table_name&.to_sym
-
-            if table_name.present?
-              attributes_considered = # then combine to get all the attrs we deal with
-                lists.map do |list|
-                  # for each list of model attrs, look up the attrs for this model
-                  list.fetch(table_name, [])
-                end.reduce(:concat)
-              missing_attributes = next_model.attribute_names - attributes_considered
-              surplus_attributes = attributes_considered - next_model.attribute_names
-
-              diff[:missing][table_name] = missing_attributes if missing_attributes.any?
-
-              diff[:surplus][table_name] = surplus_attributes if surplus_attributes.any?
-            end
+        DfE::Analytics.all_entities_in_application
+          .reduce({ missing: {}, surplus: {} }) do |diff, entity|
+            attributes_considered = lists.map do |list|
+              # for each list of model attrs, look up the attrs for this model
+              list.fetch(entity, [])
+            end.reduce(:concat)
+
+            model = DfE::Analytics.model_for_entity(entity)
+
+            missing_attributes = model.attribute_names - attributes_considered
+            surplus_attributes = attributes_considered - model.attribute_names
+
+            diff[:missing][entity] = missing_attributes if missing_attributes.any?
+
+            diff[:surplus][entity] = surplus_attributes if surplus_attributes.any?
```

Comment:
I think if you're going to be refactoring this bit, you may as well reduce the complexity here too. Maybe something like (note, these are untested sketches): 

```ruby
      def entity_attributes_from_lists(entity, lists)
        lists.map do |list|
          # for each list of model attrs, look up the attrs for this model
          list.fetch(entity, [])
        end.reduce(:concat)
      end

      def self.entities_with_missing_attributes_from_lists(*lists)
        DfE::Analytics.all_entities_in_application
          .reduce({}) do |entities_missing_attributes, entity|
            attributes_considered = entity_attributes_from_lists(entity, lists)
            model = DfE::Analytics.model_for_entity(entity)
            entities_missing_attributes[entity] = model.attribute_names - attributes_considered
          end.reject { |_entity, attributes| attributes.empty? }
      end

      def self.entities_with_surplus_attributes_from_lists(*lists)
        DfE::Analytics.all_entities_in_application
          .reduce({}) do |entities_surplus_attributes, entity|
            attributes_considered = entity_attributes_from_lists(entity, lists)
            model = DfE::Analytics.model_for_entity(entity)
            entities_surplus_attributes[entity] = model.attribute_names - attributes_considered
          end.reject { |_entity, attributes| attributes.empty? }
      end
```